server/API/REST.php

```diff
@@ -266,89 +266,122 @@ function list_lectures($vars) {
-			$query .= ' AND start_ts <= :endDate';
-		}
+				if (!$stmt->execute()) {
+					throw new PDOException($stmt->errorInfo()[2]);
+				}
 
-		// Get list of lectures, ordered
-		$query .= ' ORDER BY start_ts, ID ASC';
-		$stmt = $pdo->prepare($query);
-		$stmt->bindValue(':courseId', $courseId, PDO::PARAM_INT);
-		$stmt->bindValue(':cancelled', LECTURE_CANCELLED, PDO::PARAM_INT);
+				$affectedRows = $stmt->rowCount();
+			}
+			else{
+				$startDateString = $startDateTime->format("Y-m-d H:i");
+				$endDateString = $endDateTime == null ? null : $endDateTime->format("Y-m-d H:i");
+				$lectures = get_lectures_by_course($courseId, $startDateString, $endDateString);
+				$timeDatas = explode(":", $_PATCH["newTime"]);
+				if(count($timeDatas) != 2 || !is_numeric($timeDatas[0]) || !is_numeric($timeDatas[1])){
+					throw new ErrorException("Wrong time format, should be hh:mm");
+				}
+				$hh = intval($timeDatas[0]);
+				$mm = intval($timeDatas[1]);
 
-		if (isset($_GET['startDate'])) {
-			$Ymd = explode('-', $_GET['startDate']);
-			$ts = mktime(0, 0, 0, intval($Ymd[1]), intval($Ymd[2]), intval($Ymd[0]));
-			$stmt->bindValue(':startDate', $ts, PDO::PARAM_INT);
-		}
-		if (isset($_GET['endDate'])) {
-			$Ymd = explode('-', $_GET['endDate']);
-			$ts = mktime(59, 59, 23, intval($Ymd[1]), intval($Ymd[2]), intval($Ymd[0]));
-			$stmt->bindValue(':endDate', $ts, PDO::PARAM_INT);
-		}
+				$pdo->beginTransaction();
 
-		if (!$stmt->execute()) {
-			throw new PDOException($stmt->errorInfo()[2]);
-		}
+				$affectedRows = 0;
+				foreach($lectures["lectures"] as $lecture){
+					$cur_start_time = new DateTime('@'.$lecture["startTS"]);
+					if(intval($cur_start_time->format("w")) != $original_weekday) continue;
+					$cur_end_time = new DateTime('@'.$lecture["endTS"]);
+					$lecture_duration = $cur_end_time->getTimestamp() - $cur_start_time->getTimestamp();
 
-		$lectures = array();
-		$l = $stmt->fetch();
-		if(!$l){
-			$ret = array(
-				'courseId' => null,
-				'courseCode' => null,
-				'courseName' => null
-			);
-		}
-		else{
-			$ret = array(
-				'courseId' => intval($l['course_id']),
-				'courseCode' => $l['courseCode'],
-				'courseName' => $l['courseName'],
-			);
-			do{
-				$lecture = array(
-					'lectureId' => intval($l['0']),
-					'startTS' => intval($l['start_ts']),
-					'endTS' => intval($l['end_ts']),
-					'online' => boolval($l['settings'] & LECTURE_REMOTE),
-					'roomName' => $l['roomName']
-				);
+					$cur_start_time->modify($sign.$difference." days");
+					$cur_start_time->setTime($hh, $mm);
+					$cur_end_time->modify($sign.$difference." days");
+					$cur_end_time->setTime($hh, $mm);
+					$cur_end_time->modify("+".$lecture_duration." seconds");
 
-				array_push($lectures, $lecture);
-			} while ($l = $stmt->fetch());
-		}
+					$stmt = $pdo->prepare("UPDATE lectures SET start_ts = ".$cur_start_time->getTimestamp().", end_ts = ".$cur_end_time->getTimestamp()." WHERE ID = ".$lecture["lectureId"].$range_conditions);
```

Comment:
You can move the prepare out of the foreach statement, and only do bindValue inside